state/store_test.go

```diff
@@ -242,6 +245,204 @@ func sliceToMap(s []int64) map[int64]bool {
+		assert.Equal(t, lastResponse, response1)
+		// check if the abci response didnt save in the abciresponses.
+		responses, err := stateStore.LoadFinalizeBlockResponse(height)
+		require.NoError(t, err, responses)
+		require.Equal(t, response1, responses)
+	}
+	b1 := state.MakeBlock(state.LastBlockHeight+1, test.MakeNTxs(state.LastBlockHeight+1, 10), new(types.Commit), nil, nil)
+	partSet, err := b1.MakePartSet(2)
+	require.NoError(t, err)
+	bs.SaveBlock(b1, partSet, &types.Commit{Height: state.LastBlockHeight + 1})
+}
+
+func TestSaveRetainHeight(t *testing.T) {
+	state, bs, callbackF, stateStore := makeStateAndBlockStore()
+	defer callbackF()
+	height := int64(10)
+	state.LastBlockHeight = height - 1
+
+	fillStore(t, height, stateStore, bs, state, nil)
+
+	pruner := sm.NewPruner(stateStore, bs, log.TestingLogger())
+
+	// We should not save a height that is 0
+	err := pruner.SetApplicationRetainHeight(0)
+	require.Error(t, err)
+
+	// We should not save a height above the blockstore's height
+	err = pruner.SetApplicationRetainHeight(11)
+	require.Error(t, err)
+
+	err = pruner.SetApplicationRetainHeight(10)
+	require.NoError(t, err)
+
+	err = pruner.SetCompanionRetainHeight(10)
+	require.NoError(t, err)
+}
+
+func TestMinRetainHeight(t *testing.T) {
+	stateDB := dbm.NewMemDB()
+	stateStore := sm.NewStore(stateDB, sm.StoreOptions{
+		DiscardABCIResponses: false,
+	})
+	pruner := sm.NewPruner(stateStore, nil, log.TestingLogger())
+	minHeight := pruner.FindMinRetainHeight()
+	require.Equal(t, minHeight, int64(0))
+
+	err := stateStore.SaveApplicationRetainHeight(10)
+	require.NoError(t, err)
+	minHeight = pruner.FindMinRetainHeight()
+	require.Equal(t, minHeight, int64(10))
+
+	err = stateStore.SaveCompanionBlockRetainHeight(11)
+	require.NoError(t, err)
+	minHeight = pruner.FindMinRetainHeight()
+	require.Equal(t, minHeight, int64(10))
+}
+
+func TestABCIResPruningStandalone(t *testing.T) {
+	stateDB := dbm.NewMemDB()
+	stateStore := sm.NewStore(stateDB, sm.StoreOptions{
+		DiscardABCIResponses: false,
+	})
+
+	responses, err := stateStore.LoadFinalizeBlockResponse(1)
+	require.Error(t, err)
+	require.Nil(t, responses)
+	// stub the abciresponses.
+	response1 := &abci.ResponseFinalizeBlock{
+		TxResults: []*abci.ExecTxResult{
+			{Code: 32, Data: []byte("Hello"), Log: "Huh?"},
+		},
+	}
+	_, bs, callbackF, stateStore := makeStateAndBlockStore()
+	defer callbackF()
+
+	for height := int64(1); height <= 10; height++ {
+		err := stateStore.SaveFinalizeBlockResponse(height, response1)
+		require.NoError(t, err)
+	}
+	pruner := sm.NewPruner(stateStore, bs, log.TestingLogger())
+
+	retainHeight := int64(2)
+	err = stateStore.SaveABCIResRetainHeight(retainHeight)
+	require.NoError(t, err)
+	abciResRetainHeight, err := stateStore.GetABCIResRetainHeight()
+	require.NoError(t, err)
+	require.Equal(t, retainHeight, abciResRetainHeight)
+	pruned := pruner.PruneABCIResToRetainHeight(0)
+	require.Equal(t, retainHeight, pruned)
```

Comment:
Should this not be:

```suggestion
	require.Equal(t, retainHeight - 1, pruned)
```

We should always be pruning up to, but just before, the retain height, should we not?

Reply:
I suppose what I'm confused about is what we _should_ be returning from `pruner.pruneABCIResToRetainHeight` - should we be returning the new retain height, or the height to which we've pruned?

Reply:
The way it is currently implemented it returns the last height **up** to which we pruned (same as retain height). However, we can change this if it is more intuitive for you the other way. The only purpose of this is not to prune the same blocks twice, so this variable can mean whatever we want.

Reply:
But we shouldn't ever be pruning the retain height - only just before the retain height, should we not?

So the return value is therefore the new retain height, and not the pruned height, since the pruned height should always be `retainHeight - 1`.

As such, do we need to even return the new retain height? All we need to know is whether it was successful or not in reaching the new retain height.

Reply:
So the naming of the variable is the source of the confusion. We do not prune the retain height. We prune up to but exclude the retain height. The variable returned here is purely for optimization purposes to avoid having the pruner pruning if the new retain height is the same as the one we already had before. In some cases we will not prune up until teh requested height (due to an error, or the evidence retain height in the blockstore). In that case we want to prune again if this 'lastHeightPruned' differs from the retain height in the database.